ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/FinalProjectNode.java

```diff
@@ -56,23 +67,54 @@
   private final Optional<Analysis.Into> into;
   private final LogicalSchema schema;
   private final ImmutableList<SelectExpression> selectExpressions;
+  private final Optional<LogicalSchema> pullQueryOutputSchema;
+  private final Optional<LogicalSchema> pullQueryIntermediateSchema;
+  private final Optional<List<ExpressionMetadata>> compiledSelectExpressions;
+  private final RewrittenAnalysis analysis;
+  private final boolean isPullQuerySelectStar;
+  private final boolean addAdditionalColumnsToIntermediateSchema;
 
   public FinalProjectNode(
       final PlanNodeId id,
       final PlanNode source,
       final List<SelectItem> selectItems,
       final Optional<Analysis.Into> into,
-      final MetaStore metaStore
+      final MetaStore metaStore,
+      final KsqlConfig ksqlConfig,
+      final RewrittenAnalysis analysis
   ) {
     super(id, source);
     this.projection = Projection.of(selectItems);
     this.into = into;
+    this.analysis = Objects.requireNonNull(analysis, "analysis");
 
     final Pair<LogicalSchema, List<SelectExpression>> result = build(metaStore);
     this.schema = result.left;
     this.selectExpressions = ImmutableList.copyOf(result.right);
+    if (analysis.isPullQuery()) {
+      this.isPullQuerySelectStar = isPullQuerySelectStar();
+      this.addAdditionalColumnsToIntermediateSchema = shouldAddAdditionalColumnsInSchema();
+      this.pullQueryOutputSchema = Optional.of(buildPullQueryOutputSchema(metaStore));
+      this.pullQueryIntermediateSchema = Optional.of(buildPullQueryIntermediateSchema());
+      this.compiledSelectExpressions = Optional.of(selectExpressions
+          .stream()
+          .map(selectExpression -> CodeGenRunner.compileExpression(
+              selectExpression.getExpression(),
+              "Select",
+              pullQueryIntermediateSchema.get(),
+              ksqlConfig,
+              metaStore
+          ))
+          .collect(ImmutableList.toImmutableList()));
+    } else {
+      this.isPullQuerySelectStar = false;
+      this.addAdditionalColumnsToIntermediateSchema = false;
+      this.pullQueryOutputSchema = Optional.empty();
+      this.pullQueryIntermediateSchema = Optional.empty();
+      this.compiledSelectExpressions = Optional.empty();
+      throwOnEmptyValueOrUnknownColumns();
```

Comment:
>  What is the correspondence between DataSource and Materialization? Why do their schemas differ? If the Materialization is a table, then shouldn't the schema of the DataSource match that of the table? I.e not have the rowtime and key in the value?
>
>    Why do the output schemas of the pull query and FinalProjectNode differ? Basically, it seems that the output schemas of persistent queries are different than that of pull queries? Moreover, the output schema of FinalProjectNode seems to have extra columns that are not in the projection. Why?

Off the top of my head (see bottom), I don't know the answers to these questions (and I'm not sure that without digging anyone on the team knows) but I think these are exactly the ones we need to answer (and document for the future! Ideally in the code...) before merging this pr. Then, we can see if we can change it to be a single code path for both persistent/pull queries.

Otherwise (if we can't merge the code paths), I think we need to have two different logical nodes since they (logically) represent different operations and construct different schemas (they can extend a common base so that it's clear which code they share and which is different). We shouldn't make the pull and persistent code more intertwined but with a nuanced difference as it will make it harder to develop them independently without context on the other.

FWIW, you can change the behavior of any logical node of the persistent queries so long as the QTT tests pass. Any old query will continue to work because the physical plans are persisted, not the logical ones, in the command topic.

---


> Moreover, the output schema of FinalProjectNode seems to have extra columns that are not in the projection.

My best guess on this one is that the rowkey-in-value/rowtime-in-value/windowstart/windowend columns are needed to handle the processing. At the moment, the way we've coded it up, most operations accept only the value and not the key (or other metadata information) - so we need to copy information from the key into the value. I've looked into changing this and I got pretty close, but the problem is that the Joiner code in streams doesn't yet have the signature that I need to fully get rid of it so I abandoned the change. I'll try to revisit it (and hopefully change that in streams if necessary) at some point.